backend/baseclass.pm

```diff
@@ -168,145 +166,133 @@ sub _write_buffered_data_to_file_handle {
+        #   Disadvantages: Requires changes on both ends of the communication. (for example: on SSH connection i realized that after a
+        #   while I start getting "bad packet length" errors)
+        # + Polling the connections (Note: This is how HTTP servers work when dealing with persistent connections)
+        #    Disadvantages: False positives
+        # + Change the keepalive packet settings
+        #   Disadvantages: TCP/IP stacks are not required to support keepalives.
+        if (fileno $fh && fileno $fh != -1) {
+            # Very high limits! On a working socket, the maximum hits per 10 seconds will be around 60.
+            # The maximum hits per 10 seconds saw on a half open socket was >100k
+            if (check_select_rate($buckets, $wait_time_limit, $hits_limit, fileno $fh)) {
+                my $console = $self->{current_console}->{testapi_console};
+                OpenQA::Exception::ConsoleReadError->throw(error => "The console '$console' is not responding (half-open socket?). Make sure the console is reachable or disable stall detection on expected disconnects with '\$console->disable_vnc_stalls', for example in case of intended machine shutdown");
             }
+        }
 
-            # if we got stalled for a long time, we assume bad hardware and report it
-            if ($self->assert_screen_last_check && $now - $self->last_screenshot > $self->screenshot_interval * 20) {
-                $self->stall_detected(1);
-                my $diff = $now - $self->last_screenshot;
-                bmwqemu::fctwarn "There is some problem with your environment, we detected a stall for $diff seconds";
-            }
 
-            my $time_to_screenshot = $self->screenshot_interval - ($now - $self->last_screenshot);
-            if ($time_to_screenshot <= 0) {
-                $self->capture_screenshot();
-                $self->last_screenshot($now);
-                $time_to_screenshot = $self->screenshot_interval;
-            }
+        unless ($self->check_socket($fh, 0)) {
+            die "huh! $fh\n";
+        }
+        # don't check for further sockets after this one as
+        # check_socket can have side effects on the sockets
+        # (e.g. console resets), so better take the next socket
+        # next time
+        last;
+    }
+}
 
-            my $time_to_next = min($time_to_screenshot, $time_to_update_request, $time_to_timeout);
-            my ($read_set, $write_set) = IO::Select->select($self->{select_read}, $self->{select_write}, undef, $time_to_next);
+sub capture_loop {
+    my ($self, $timeout, $starttime) = @_;
+    # Time slot buckets
+    my $buckets         = {};
+    my $wait_time_limit = $bmwqemu::vars{_CHKSEL_RATE_WAIT_TIME} // 30;
+    my $hits_limit      = $bmwqemu::vars{_CHKSEL_RATE_HITS}      // 15_000;
+    $self->do_capture($timeout, $starttime, $buckets, $wait_time_limit, $hits_limit) while 1;
```

Comment:
Is this loop supposed to be interrupted by a `die`?

Reply:
reverted